src/main/scala/io/iohk/ethereum/consensus/ethash/RestrictedEthashSigner.scala

```diff
@@ -0,0 +1,33 @@
+package io.iohk.ethereum.consensus.ethash
+
+import akka.util.ByteString
+import io.iohk.ethereum.crypto
+import io.iohk.ethereum.crypto.ECDSASignature
+import io.iohk.ethereum.domain.BlockHeader
+import io.iohk.ethereum.domain.BlockHeader.getEncodedWithoutNonce
+import org.bouncycastle.crypto.AsymmetricCipherKeyPair
+
+object RestrictedEthashSigner {
+
+  def validateSignature(blockHeader: BlockHeader, allowedMiners: Set[ByteString]): Boolean = {
+    val signature = blockHeader.extraData.takeRight(ECDSASignature.EncodedLength)
+    val blockHeaderWithoutSig = blockHeader.dropRightNExtraDataBytes(ECDSASignature.EncodedLength)
+    val encodedBlockHeader = getEncodedWithoutNonce(blockHeaderWithoutSig)
+    val headerHash = crypto.kec256(encodedBlockHeader)
+    val maybePubKey = for {
+      sig <- ECDSASignature.fromBytes(signature)
+      pubKeyBytes <- sig.publicKey(headerHash)
+    } yield ByteString.fromArrayUnsafe(pubKeyBytes)
+
+    maybePubKey.exists(allowedMiners.contains)
+  }
+
+  def signHeader(blockHeader: BlockHeader, keyPair: AsymmetricCipherKeyPair): BlockHeader = {
+    val encoded = getEncodedWithoutNonce(blockHeader)
```

Comment:
Why without the nonce and mixHash?

Reply:
Becouse those are modified by mining i.e mining is always done on whole block data except `nonce` and `mixHash`, and correct values are of `nonce` and `mixHash` are provided by mining result.

Reply:
Oh, because we are obtaining the correct nonce and mixHash after this... we could have done the other way around, but I see no problem between either alternative, it will always have been generated by a whitelisted miner, aka a trusted one